File: web/src/features/attendance/services/attendanceService.js
import axiosInstance from '../../../api/axiosInstance';

const BASE = '/api/v1/attendance';

export const getAttendanceByCourseAndDate = (courseId, date) =>
  axiosInstance.get(`${BASE}/course/${courseId}`, { params: { date } }).then((r) => r.data);

export const markAttendanceBatch = (data) =>
  axiosInstance.post(BASE, data).then((r) => r.data);

export const getMyAttendanceSummary = () =>
  axiosInstance.get(`${BASE}/me/summary`).then((r) => r.data);

export const getStudentAttendanceSummary = (studentId) =>
  axiosInstance.get(`${BASE}/student/${studentId}/summary`).then((r) => r.data);

export const getClassCourseOverview = (classId, courseId) =>
  axiosInstance.get(`${BASE}/class/${classId}/course/${courseId}/overview`).then((r) => r.data);

export const getOverviewByClassStructure = (classStructureId, courseId) =>
  axiosInstance.get(`${BASE}/class-structure/${classStructureId}/course/${courseId}/overview`).then((r) => r.data);

export const getAttendanceByCourseAndClassStructure = (courseId, classStructureId, date) =>
  axiosInstance.get(`${BASE}/course/${courseId}/class-structure/${classStructureId}`, { params: { date } }).then((r) => r.data);

export const getAttendanceByDateRange = (courseId, startDate, endDate) =>
  axiosInstance.get(`${BASE}/course/${courseId}/range`, { params: { startDate, endDate } }).then((r) => r.data);

File: api/collegeportal/src/main/java/com/collegeportal/modules/attendance/service/AttendanceService.java
package com.collegeportal.modules.attendance.service;

import com.collegeportal.modules.attendance.dto.request.AttendanceBatchRequestDTO;
import com.collegeportal.modules.attendance.dto.request.AttendanceRequestDTO;
import com.collegeportal.modules.attendance.dto.request.AttendanceUpdateRequestDTO;
import com.collegeportal.modules.attendance.dto.response.AttendanceResponseDTO;
import com.collegeportal.modules.attendance.dto.response.AttendanceSummaryDTO;
import com.collegeportal.modules.attendance.dto.response.StudentAttendanceOverviewDTO;
import com.collegeportal.shared.dto.PageResponseDTO;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;

public interface AttendanceService {

    AttendanceResponseDTO markAttendance(AttendanceRequestDTO request);

    AttendanceResponseDTO updateAttendance(Long id, AttendanceUpdateRequestDTO request);

    List<AttendanceResponseDTO> markAttendanceBatch(List<AttendanceBatchRequestDTO> requests);

    List<AttendanceResponseDTO> getAttendanceByCourseAndDate(Long courseId, LocalDate date);

    List<AttendanceResponseDTO> getAttendanceByCourseClassAndDate(Long courseId, Long classId, LocalDate date);

    PageResponseDTO<AttendanceResponseDTO> getMyAttendance(Pageable pageable);

    List<AttendanceSummaryDTO> getMyAttendanceSummary();

    List<AttendanceSummaryDTO> getStudentAttendanceSummary(Long studentId);

    /** Returns one row per student in the class for the given course — used by admin/faculty overview table */
    List<StudentAttendanceOverviewDTO> getClassCourseOverview(Long classId, Long courseId);

    /** Same overview but keyed by ClassStructure (admin courses flow) */
    List<StudentAttendanceOverviewDTO> getOverviewByClassStructure(Long classStructureId, Long courseId);

    /** Attendance records for a course on a date, scoped to a ClassStructure */
    List<AttendanceResponseDTO> getAttendanceByCourseAndClassStructure(Long courseId, Long classStructureId, LocalDate date);

    /** Attendance records for a course within a date range, scoped to the current faculty */
    List<AttendanceResponseDTO> getAttendanceByCourseAndDateRange(Long courseId, LocalDate startDate, LocalDate endDate);
}
